### PR TITLE
[NEW] Chat Details

### LINKED ISSUE #835: [NEW] Display Chat Room Information
**Feature**
Add a room option which displays in a _small pop-up window_ the information about a public chat room (such as its topic, last announcement, creator, etc.)

**Willing to work on this**
Yes

### PR DESCRIPTION
**Pull request by samrmur**
@RocketChat/android

Closes #835 

**PR Info**
Adds the capability of viewing a chat room's information. This works for both subscribed and unsubscribed rooms.

Information that can be viewed:
- Channel Name
- Topic
- Description
- Announcement

Note: Could not find all icons that were shown in mock. Can implement those if available.

**Screenshot**

### CODE REVIEW

**Comment by rafaelks** [comment]
@samrmur Amazing, thank you! @filipedelimabrito is going to review your PR! 👍

**Comment by rafaelks** [comment]
The app is crashing for me on the first start... does it work fine for you?

**Comment by samrmur** [comment]
Application seems to be working fine for me. I'll update this PR and fix the most recent conflict shortly.

**Comment by samrmur** [comment]
@rafaelks updated

**Comment by rafaelks** [comment]
@samrmur Do you mind updating the app again? We'll make sure to review your PR asap.

**Comment by rafaelks** [comment]
App still doesn't open for me... crashes right after I open.

File: app/src/main/java/chat/rocket/android/chatdetails/di/ChatDetailsFragmentModule.kt [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import chat.rocket.android.dagger.scope.PerFragment
+import chat.rocket.android.db.ChatRoomDao
+import chat.rocket.android.db.DatabaseManager
+import dagger.Module
+import dagger.Provides
+import kotlinx.coroutines.experimental.Job
+
+@Module
+class ChatDetailsFragmentModule {
+    @Provides
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/chatdetails/di/ChatDetailsFragmentModule.kt:17]: Add a space here.

File: app/src/main/java/chat/rocket/android/chatdetails/di/ChatDetailsFragmentProvider.kt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package chat.rocket.android.chatdetails.di
+
+import chat.rocket.android.chatdetails.ui.ChatDetailsFragment
+import chat.rocket.android.dagger.scope.PerFragment
+import dagger.Module
+import dagger.android.ContributesAndroidInjector
+
+@Module
+abstract class ChatDetailsFragmentProvider {
+    @ContributesAndroidInjector(modules = [ChatDetailsFragmentModule::class])
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/chatdetails/di/ChatDetailsFragmentProvider.kt:11]: Add a space here.

File: app/src/main/java/chat/rocket/android/chatdetails/di/ChatDetailsModule.kt [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+import chat.rocket.android.chatdetails.presentation.ChatDetailsNavigator
+import chat.rocket.android.chatdetails.ui.ChatDetailsActivity
+import chat.rocket.android.dagger.scope.PerActivity
+import dagger.Module
+import dagger.Provides
+
+@Module
+class ChatDetailsModule {
+    @Provides
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/chatdetails/di/ChatDetailsModule.kt:12]: Add a space here.

File: app/src/main/java/chat/rocket/android/chatdetails/presentation/ChatDetailsNavigator.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import chat.rocket.android.chatdetails.ui.ChatDetailsActivity
+import chat.rocket.android.favoritemessages.ui.TAG_FAVORITE_MESSAGES_FRAGMENT
+import chat.rocket.android.files.ui.TAG_FILES_FRAGMENT
+import chat.rocket.android.members.ui.TAG_MEMBERS_FRAGMENT
+import chat.rocket.android.mentions.ui.TAG_MENTIONS_FRAGMENT
+import chat.rocket.android.pinnedmessages.ui.TAG_PINNED_MESSAGES_FRAGMENT
+import chat.rocket.android.util.extensions.addFragmentBackStack
+
+class ChatDetailsNavigator(internal val activity: ChatDetailsActivity) {
+    fun toMembersList(chatRoomId: String) {
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/chatdetails/presentation/ChatDetailsNavigator.kt:14]: Add a space here.

File: app/src/main/java/chat/rocket/android/chatdetails/presentation/ChatDetailsPresenter.kt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    factory: ConnectionManagerFactory
+) {
+    private val currentServer = serverInteractor.get()!!
+    private val manager = factory.create(currentServer)
+    private val client = manager.client
+
+    fun getDetails(chatRoomId: String, chatRoomType: String) {
+        launchUI(strategy) {
+            try {
+                view.displayDetails(roomToChatDetails(client.getInfo(chatRoomId, null, roomTypeOf(chatRoomType))))
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/chatdetails/presentation/ChatDetailsPresenter.kt]: Better to use the `retryIO` here `client.getInfo(chatRoomId, null, roomTypeOf(chatRoomType))`

File: app/src/main/java/chat/rocket/android/chatdetails/ui/ChatDetailsAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+class ChatDetailsAdapter(private val context: Context): RecyclerView.Adapter<ChatDetailsAdapter.ViewHolder>() {
+    private val options: MutableList<Option> = ArrayList()
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder = ViewHolder(parent.inflate(R.layout.item_detail_option))
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) = holder.bind(options[position])
+
+    override fun getItemCount(): Int = options.size
+
+    inner class ViewHolder(itemView: View): RecyclerView.ViewHolder(itemView) {
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/chatdetails/ui/ChatDetailsAdapter.kt]: try avoiding the use of inner class here.

File: app/src/main/java/chat/rocket/android/chatdetails/ui/ChatDetailsAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        fun bind(option: Option) {
+            bindName(option, itemView.name)
+            bindIcon(option, itemView.icon)
+            itemView.setOnClickListener { option.listener() }
+        }
+    }
+
+    inner class Option(val name: String, val icon: Int, val listener: () -> Unit)
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/chatdetails/ui/ChatDetailsAdapter.kt]: Here too.

File: app/src/main/java/chat/rocket/android/chatdetails/ui/ChatDetailsFragment.kt [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+                                it,
+                                resources.getDimensionPixelSize(R.dimen.divider_item_decorator_bound_start),
+                                resources.getDimensionPixelSize(R.dimen.divider_item_decorator_bound_end)
+                        )
+                    )
+                    itemAnimator = DefaultItemAnimator()
+                    isNestedScrollingEnabled = false
+                }
+            }
+            addOptions(adapter!!)
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/chatdetails/ui/ChatDetailsFragment.kt]: `!!` 🛑

File: app/src/main/java/chat/rocket/android/chatdetails/viewmodel/ChatDetailsViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.Transformations
+import androidx.lifecycle.ViewModel
+import chat.rocket.android.chatdetails.domain.ChatDetails
+import chat.rocket.android.db.ChatRoomDao
+
+class ChatDetailsViewModel(private val chatRoomDao: ChatRoomDao): ViewModel() {
+    fun getDetails(chatRoomId: String): LiveData<ChatDetails> {
```
Comment (philipbrito) [inline app/src/main/java/chat/rocket/android/chatdetails/viewmodel/ChatDetailsViewModel.kt:12]: Add a space here.

**Pull request merged**